[WIP] Add avl tests

TheNeikos opened the pull request:
This pull request adds and fixes the avl tests

matthiasbeyer committed:
Added testing file for AVL implementation

matthiasbeyer committed:
Added allocation test for avl test

matthiasbeyer committed:
Added test for testing allocation and deleting for AVL

matthiasbeyer committed:
Added test for adding data to the avl tree

matthiasbeyer committed:
Added test for adding and destroying afterwards

matthiasbeyer committed:
Added test for adding multiple elements

matthiasbeyer committed:
Added test for adding multiple and destroying AVL afterwards

matthiasbeyer committed:
Added test for colliding inserts

matthiasbeyer committed:
Added test for finding a single element

matthiasbeyer committed:
Added test for finding multiple elements

matthiasbeyer committed:
[ci skip] Fixup: test_avl_add: The ptrs to the l/r next element should be NULL, actually

TheNeikos committed:
Fixed compiler errors about a const modified

TheNeikos committed:
Fixes a bug in the avl_find test

I also updated the documentation of avl.h as
@neithernut forgot to update it.

TheNeikos committed:
Changed tests to conform to master

TheNeikos committed:
Fixed an inverted condition

TheNeikos committed:
Fixed occurences of NULL and removed internal

matthiasbeyer commented:
The build fails, so I cannot merge - any futher explanation on this?

matthiasbeyer commented on tests/set_cfg.c:
```diff
@@ -3,7 +3,7 @@
 
 
-static rs_hash hashf(void const*);
+static rs_hash hashf(void*);
```
Why was the `const` keyword removed?

TheNeikos replied:
I will have to look, it might come from one of the previous commits(made by you)! Good catch.

matthiasbeyer commented on tests/avl/avl_tests.c:
```diff
@@ -0,0 +1,202 @@
+#include "avl.h"
+#include "set_cfg.h"
+
+START_TEST (test_avl_alloc) {
```
May be removed as you only test the `malloc()` function here. (And `malloc()` does not fill the memory with `0`, so you probably end up with a failing test here anyways!

matthiasbeyer commented on tests/avl/avl_tests.c:
```diff
@@ -0,0 +1,202 @@
+    struct avl* avl = malloc(sizeof(*avl));
+
+    ck_assert(avl != NULL);
+    ck_assert(avl->root == NULL);
```
Cannot be guaranteed, as `malloc()` does not fill the memory with zero here. Use `calloc`.

matthiasbeyer commented on tests/avl/avl_tests.c:
```diff
@@ -0,0 +1,202 @@
+    int data = 1;
+    rs_hash hash = 1;
+
+    struct avl_el* element = avl_add(avl, hash, &data, &cfg_int);
```
Why storing the element here, if not checking for a `NULL` ptr return?

matthiasbeyer commented:
Basically a good PR! Some things are left to do, but I'll appreciate your work here, thanks!

matthiasbeyer commented:
Btw: 

> I'm not a friend on fiddeling with internas of data structures.
> The tests should only use the interface of a datastructure!

:-P

TheNeikos commented:
Heh, yes sorry forgot to add the [wip] tag! Thanks for the review, gonna work on that.

neithernut commented on tests/avl/avl_tests.c:
```diff
@@ -0,0 +1,202 @@
+
+        ck_assert_int_eq(element->hash, hash[i]);
+    }
+    ck_assert_int_eq(avl_node_cnt(avl), 10);
```
`avl_node_cnt()` actually takes a node as argument, but you are passing the tree.
If you change this to `avl_node_cnt(avl->root)` the test passes and issue #157 can be closed.
However, I experience a segfault now in another test (I'm working on that issue).

TheNeikos replied:
Ah! Thanks, that does indeed explain it.

neithernut commented on tests/avl/avl_tests.c:
```diff
@@ -0,0 +1,202 @@
+END_TEST
+
+START_TEST (test_avl_find_single) {
+    struct avl* avl = malloc(sizeof(*avl));
```
`malloc()` does not null the memory, so `avl->root` is some random value.
That caused the crash described earlier.

neithernut commented:
I'd like to see this branch rebased on newest master as soon as the issues are fixed (and maybe after PR #162 and PR #163) are merged.
I will then proceed tracking down the issues left. (some tests are still failing)

TheNeikos committed:
Fixed tests to conform to latest api

TheNeikos commented:
Hm, I rebased on master, yet it shows me @matthiasbeyer's commits... any suggestions for that?

matthiasbeyer commented:
That's because I actually wrote these commits. If you look at `git show` for one of the commits, you will see that you are the committer of these, but I am the author. Everything is fine with this!

TheNeikos commented:
Alright then.

TheNeikos commented:
Once the Travis build is done one can see that one of the tests seg faults.

The test `avl_insert_multiple` also fails because `avl_node_cnt(avl->root)` returns 2 instead of the expected 10.

[avl_insert_collisions](https://github.com/matthiasbeyer/libreset/pull/156/files#diff-89873f4aca91ecfeca063a09e8b8bd0fR92) seg faults. I am not sure why. `gdb` tells me it segfaults at [avl.c#268](https://github.com/matthiasbeyer/libreset/blob/master/src/libreset/avl.c#L268) , which is [avl.c#272](https://github.com/TheNeikos/libreset/blob/add-avl_test/src/libreset/avl.c#L272) in this PR.

matthiasbeyer commented:
Is this a bug in `avl_insert_multiple()` or in other code? If the first, please file an issue refering to this PR and try to fix it in a seperate PR.

I will try to trace down the bug in the `test_avl_insert_collisions` test function ASAP as I'm, done with my current stuff.

matthiasbeyer committed:
Fixup: Style violation

matthiasbeyer committed:
Fix: SEGFAULT in tests_avl_insert_multiple_destroy

The use of `malloc()` caused to allocate an avl structure with
initialized `root` pointer member. At some point when iterating through
this tree of non-existing memory, it segfauls. Allocating the avl
structure with a NULL root node causes the code to allocate _real_
nodes, which _can_ be accessed.

matthiasbeyer committed:
Fix: Possible segfaults, allocate the avl structure with NULLed memory

matthiasbeyer committed:
Remove: test_avl_alloc: We do not test the calloc() function, it has to work

TheNeikos committed:
Added a cast to (int*)

TheNeikos commented:
Okay, thanks for those! Wouldn't have thought of malloc causing problems... (Noob error?)

However, another problem still persists with `avl_insert_multiple` not adding all the nodes that are needed. I will open an issue for that.

matthiasbeyer commented:
> (Noob error?)

Not really, but also an "an expert wouldn't have done this wrong" error.

matthiasbeyer commented:
@TheNeikos please rebase. #190 fixed the issue which was also reported in #186 . I tested it, the issue was fixed, but there are others. After you rebased, notify me for more fixes.

neithernut commented on tests/avl/avl_tests.c:
```diff
@@ -0,0 +1,180 @@
+
+    for (i = 0; i < 10; i++) {
+        ck_assert(&data[i] == avl_find(avl, hash[i], &data[i], &cfg_int));
+        free(elements[i]);
```
No single element of the array `elements` ever gets a value assigned.

TheNeikos replied:
Ah yes, in a hurry to get it working with the new api (since avl_insert now returns an int) I forgot to remove elements and the free.

neithernut commented:
Please rebase on current master.

TheNeikos committed:
Removed uneeded elements array

TheNeikos committed:
Replaced a free with an avl_destroy

TheNeikos committed:
Added a missing const

TheNeikos commented:
The output of Travis CI doesn't make any sense to me... [This part specifically](https://travis-ci.org/matthiasbeyer/libreset/builds/35955331#L1351) Apparently `1 != 1` ...

On my machine `ctest -VV .` does not report any problems what so ever, I am confused.

neithernut commented:
Again, nothing comming up on this machine (all tests passing).
Interestingly, I also do not get any double-frees or corruptions.

TheNeikos commented:
Okay, with this all the tests pass on my local machine. I do not get any double frees either.

TheNeikos committed:
Changed the avl.h include path to the real on

TheNeikos committed:
Removed the int_eq and replaced it with plain comp

matthiasbeyer commented:
I guess you can remove e508123893305c81981445d0d2f221e6e754b79d

---

Bra, told you bullshit. You can replace `a == NULL` with `!a`.

matthiasbeyer commented:
ACK.
